### PR TITLE
Deduplicate fields and outputs

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

This is a early draft. Not sure if this is going to work out.

Rough outline:

- Removes `fields` from `AnalyzedRelation`. There are only `outputs`

- Changes `Field` to `ScopedSymbol`, which is only added when a relation
  introduces a new scope. (E.g. in SELECT t1.x, t2.x FROM t AS t1, t AS
  t2, it is necessary to distinguish between t1.x and t2.x, so `AS t1`
  needs to introduce a scope)

- Adds `AliasSymbol` to preserve the `AS <xy>` information

- Changes the `LogicalPlanner` to only process the clauses of a relation that
  are really there. (Only `QueriedSelectRelation` can contain WHERE, GROUP BY,
  HAVING, etc...)

- Merged `QueriedSelectRelation` and `MultiSourceSelect`. Also changes
  `QualifiedName` usages to `RelationName` - so that everything is unified to
  `RelationName` after the Analyzer.

- ??

- Profit

Things that can be followed up on in later PRs:

- [ ] Remove WhereClause/HavingClause usage in the AnalyzedRelation and replace it with simple Symbols. It is trappy in the current state because the `Filter` operator drops `partition` information. (A non issue right now, but stab us in the back later)
- [ ] Remove `SplitPoints` and eager `toCollect/outputs` propagation in the `PlanBuilder`
- [ ] Subscript handling can likely be cleaned up more. (And optimized)
- [ ] Proper SymbolPrinting support for window functions and re-create COUNT(DISTINCT)
- [ ] Get rid of `ExplainLeaf` representation and use SymbolPrinter everywhere instead.
- [ ] Maybe introduce some common base class for `ScopedSymbol` and `Reference`, or make `Reference` extend `ScopedSymbol`, so that in many places we don't have to handle them both. Or maybe go the other route and have `ScopedSymbol` everywhere except the very last moment when building the execution plan.

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mfussenegger** [comment]
@seut @kovrus Maybe you can have a look at this. It is far from finished but I want to see if this is going in the right direction.

A lot of analyzer tests are failing (expected), some integration tests are failing (some expected, due to not having adapted the subscript handling and some wrong splitPoints handling in the planner)

I previously had tried to do things that this PR contains in smaller steps, but it looks like doing all at once might be easier to pull off.

The main idea is to de-duplicate `outputs` and `fields` by splitting `Field` into a `ScopedSymbol` and into a `AliasSymbol`. Some relations no longer introduce a new scope (because it is not necessary). I think so far it will only be necessary for Union and Alias relations to have a separate scope.

**Comment by seut** [comment]
@mfussenegger Looks really nice. The only issue to discuss I've encountered is that I think that semantically a view also always introduce a scope (https://github.com/crate/crate/pull/9684/commits/7177739ca440886988c22a27aef0447dc2f717d8#diff-44a5c7e377fd99d284e7a5ed622c2d06R52) and I was thinking of always creating that scope inside the analyzer/planner and optimize it later if possible. Maybe this is also your idea and I mis-read it ;) I think when refactoring the subscript handling, a scoped view is maybe even required.

**Comment by kovrus** [comment]
looks nice! 👍  

can't really provide an input on the scoped view.

File: sql/src/test/java/io/crate/analyze/relations/ExcludedScopedSymbolProviderTest.java [hunk 1]
```diff
@@ -23,36 +23,34 @@
...
 import org.junit.Test;
-import org.mockito.Mockito;
 
 import static io.crate.testing.SymbolMatchers.isField;
 import static io.crate.testing.SymbolMatchers.isLiteral;
-import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertThat;
 
-public class ExcludedFieldProviderTest {
+public class ExcludedScopedSymbolProviderTest {
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/test/java/io/crate/analyze/relations/ExcludedScopedSymbolProviderTest.java]: Yep, Intellij rename refactoring was too aggressive :(

File: sql/src/main/java/io/crate/analyze/relations/UnionSelect.java [hunk 2]
```diff
@@ -22,46 +22,36 @@
...
-        this.name = left.getQualifiedName();
-
-        List<Field> fieldsFromLeft = left.fields();
-        fields = new Fields(fieldsFromLeft.size());
-        for (Field field : fieldsFromLeft) {
-            // Creating a field that points to the field of the left relation isn't 100% accurate.
-            // We're pointing to *two* symbols (both left AND right).
-            // We could either use a `InputColumn` to do that (by pointing to a position) - (but might be confusing to have InputColumns in the analysis already)
-            // Or introduce a `UnionSymbol` or `UnionField` which would take two symbols it is pointing to
-            // Since this currently has no effect we go with the left symbol until there is a good reason to change it.
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/relations/UnionSelect.java]: Not sure - the pointer stuff is gone. I could add a comment that due to SQL semantics it uses the name of the left relation or so?

[the remark this answers is not in the document]
Comment (seut) [reply sql/src/main/java/io/crate/analyze/relations/UnionSelect.java]: Hm right, fine for me without a comment as well.

[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/relations/UnionSelect.java]: Added a comment anyways

**Pull request merged**